ql/src/java/org/apache/hadoop/hive/ql/exec/repl/bootstrap/load/LoadConstraint.java

```diff
@@ -0,0 +1,119 @@
+import org.apache.hadoop.hive.ql.exec.Task;
+import org.apache.hadoop.hive.ql.exec.repl.bootstrap.events.ConstraintEvent;
+import org.apache.hadoop.hive.ql.exec.repl.bootstrap.load.util.Context;
+import org.apache.hadoop.hive.ql.parse.EximUtil;
+import org.apache.hadoop.hive.ql.parse.SemanticException;
+import org.apache.hadoop.hive.ql.parse.repl.DumpType;
+import org.apache.hadoop.hive.ql.parse.repl.load.DumpMetaData;
+import org.apache.hadoop.hive.ql.parse.repl.load.message.AddForeignKeyHandler;
+import org.apache.hadoop.hive.ql.parse.repl.load.message.AddNotNullConstraintHandler;
+import org.apache.hadoop.hive.ql.parse.repl.load.message.AddPrimaryKeyHandler;
+import org.apache.hadoop.hive.ql.parse.repl.load.message.AddUniqueConstraintHandler;
+import org.apache.hadoop.hive.ql.parse.repl.load.message.MessageHandler;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.hadoop.hive.ql.parse.BaseSemanticAnalyzer.stripQuotes;
+
+public class LoadConstraint {
+  private static final Logger LOG = LoggerFactory.getLogger(LoadFunction.class);
+  private Context context;
+  private final ConstraintEvent event;
+  private final String dbNameToLoadIn;
+  private final TaskTracker tracker;
+
+  public LoadConstraint(Context context, ConstraintEvent event, String dbNameToLoadIn,
+      TaskTracker existingTracker) {
+    this.context = context;
+    this.event = event;
+    this.dbNameToLoadIn = dbNameToLoadIn;
+    this.tracker = new TaskTracker(existingTracker);
+  }
+
+  public TaskTracker tasks() throws IOException, SemanticException {
+    URI fromURI = EximUtil
+        .getValidatedURI(context.hiveConf, stripQuotes(event.rootDir().toUri().toString()));
+    Path fromPath = new Path(fromURI.getScheme(), fromURI.getAuthority(), fromURI.getPath());
+
+    try {
+      FileSystem fs = FileSystem.get(fromPath.toUri(), context.hiveConf);
+      JSONObject json = new JSONObject(EximUtil.readAsString(fs, fromPath));
+      String pksString = json.getString("pks");
+      String fksString = json.getString("fks");
+      String uksString = json.getString("uks");
+      String nnsString = json.getString("nns");
+      List<Task<? extends Serializable>> tasks = new ArrayList<Task<? extends Serializable>>();
+
+      AddPrimaryKeyHandler pkHandler = new AddPrimaryKeyHandler();
+      DumpMetaData pkDumpMetaData = new DumpMetaData(fromPath, DumpType.EVENT_ADD_PRIMARYKEY, Long.MAX_VALUE, Long.MAX_VALUE, null,
+          context.hiveConf);
+      pkDumpMetaData.setPayload(pksString);
+      tasks.addAll(pkHandler.handle(
+          new MessageHandler.Context(
+              dbNameToLoadIn, null, fromPath.toString(), null, pkDumpMetaData, context.hiveConf,
+              context.hiveDb, null, LOG)));
+
+      AddForeignKeyHandler fkHandler = new AddForeignKeyHandler();
```

Comment:
It is possible that foreign key table is missing in bootstrap dump but fk constraint is added. This can happen when table is added after fetching table names and foreign key is set based on this table. This leads to InvalidObjectException when try to add this foreign key.

Reply:
How about adding an optional flag in AlterTableDesc? So AlterTableTask won't fail if create fk fail.

Reply:
Yeah. This is fine. Anyways, the catch-up incremental repl will add fk again after creating foreign table.

Reply:
Ok, catch exception if it is during replication.

Reply:
Yeah, this is fine.